src/builtin.jq

```diff
@@ -273,3 +292,136 @@ def JOIN($idx; stream; idx_expr; join_expr):
   stream | [., $idx[idx_expr]] | join_expr;
 def IN(s): any(s == .; .);
 def IN(src; s): any(src == s; .);
+
+def _try_finally(e; h; f):
+  . as $dot |
+  unwinding |
+  if .==false then $dot|try(e; h)
+  else f
+  end;
+
+# Default I/O policy evaluator.
+#
+# Input must be of form {io_request:<request>,io_policy:<policy>},
+# where <request> is:
+#
+#  {kind:"<kind>",name:"<path/command>"}
+#  {kind:"<kind>",name:"<path/command>",mode:"<file mode>"}
+#
+# <policy> is one of:
+#
+#  [<policy>,...]
+#  {kinds:{"<kind>":{"<name>":<policy>}}}
+#  {kind:"<kind>",name:"<name>",mode:[<mode>,...]}
+#  {kind:"<kind>",prefix:"<prefix>",mode:[<mode>,...]}
+#
+# <kind> is one of "file", "popen", "system", or "spawn"
+# <name> is the a file path or command
+# <prefix> is a string prefix of a file path ('/' gets no special treatment)
+def default_io_policy_check:
+  select(.!=null) |
+  .io_request as $req |
+
+  def check_file:
+    (.kind==$req.kind) and
+    (.mode|select(.!=null)|any(. == ($req.mode))) and
+    ((.name == $req.name) or
+     ((.prefix|type) == "string" and .prefix as $prefix | ($req.name|startswith($prefix))));
+
+  def check_cmd:
+    (.kind==$req.kind) and
+    (.mode|select(.!=null)|any(. == ($req.mode))) and
+    .name == $req.name;
+
+  def check:
+    if $req.kind=="file" then check_file
+    elif $req.kind=="popen" then check_cmd
+    elif $req.kind=="system" then check_cmd
+    # Note that file actions for spawn are checked separately, though
+    # we could also check them here.
+    elif $req.kind=="spawn" then check_cmd
+    else false end;
+
+  .io_policy as $policy |
+  if $policy == true then
+    true
+  elif ($policy|type) == "array" then
+    $policy|any(check)
+  elif ($policy|type) == "object" then
+    ($policy.kinds[$req.kind//""][$req.name//""]|(.kind=$req.kind)|(.name=$req.name)|check) or ($policy|check)
+  else
+    false
+  end;
+
+def coeval(program; args; options): [program, ., args, options] | coeval;
+def eval(program; args; options):
+  label $out |
+  coeval(program; args; options) |
+  repeat(try fhread catch if .=="EOF" then break $out else error end);
+def eval: . as $dot | null | eval($dot; {}; {});
+def coeval: . as $program | null | coeval($program; {}; {});
+
+def coexp(cexp):
+  null |
+  cocreate as $child |
+  if ($child == false) or $child == true
+  then null | cexp | cooutput # child;  call and output (which always backtracks!)
```

Comment:
Here and in other similar places, why not do the `. as $dot` thing and pass the input to the coroutine?
```
then $dot | cexp | cooutput
```

Reply:
The idea is three-fold:

1. avoid accidentally holding on to one more reference of `coexp/1`'s input that the expression might not need,
2. make it possible to send new inputs to co-expressions by writing to their file handles,
3. avoid confusion over whether some input is the initial input or one written later.

I admit this requires more thought!  Maybe (3) doesn't matter, then neither does (1).  Or, perhaps we should say that the co-expression doesn't start until you explicitly write a value to it.

Also, take a look at `coinit()` and `coinput_cb()` in `src/execute.c`, and `cowrite()` in src/builtin.c`...  Notice anything funny?  Right!  There's a one-value buffer only.  So if you write N values to a co-routine before extracting any outputs from it, then it will only get the last value written the next time it tries to read a value using `input`.  I was... lazy, for sure, but also, this requires some thought as well.